lib/faraday_connection_pool.rb

```diff
@@ -0,0 +1,71 @@
+require 'monitor'
+require 'connection_pool'
+
+module FaradayConnectionPool
+
+  class Configuration
+    attr_writer :size
+    attr_writer :timeout
+
+    def size
+      @size ||= 5
+    end
+
+    def timeout
+      @timeout ||= 5
+    end
+  end
+
+  def self.configure
+    yield configuration if block_given?
+  end
+
+  def self.configuration
+    @configuration ||= FaradayConnectionPool::Configuration.new
+  end
+
+  class Adapter
+    class NetHttpPooled < Faraday::Adapter::NetHttp
+      extend MonitorMixin
+      @@connection_pools = {}
+
+      def call(env)
+        env[:request_headers] ||= {}
+        env[:request_headers]['Connection'] = 'Keep-Alive' unless env[:request_headers]['Connection']
+        super
+      end
+
+      def with_net_http_connection(env)
+        connection_pool_for(env).with do |connection|
+          yield connection
+         end
+      end
+
+      def self.purge_connection_pools
+        synchronize do
+          @@connection_pools = {}
+        end
+      end
+
+      private
+      def connection_pool_for(env)
+        self.class.synchronize do
+          @@connection_pools[pool_key(env)] ||=
+            ConnectionPool.new(:size => FaradayConnectionPool.configuration.size,
+                               :timeout => FaradayConnectionPool.configuration.timeout) { net_http_connection(env) }
```

Comment:
I think you should have the keepalive configurable. Net::HTTP has a really low default keep-alive time ([2s](http://ruby-doc.org/stdlib-2.1.0/libdoc/net/http/rdoc/Net/HTTP.html#keep_alive_timeout)) and I think in most cases it makes sense to have a much longer value (like, 60s)